Internet midi respects start and end times

JJRcop opened the pull request:
I was looking through [my rounds](https://sb.atlantaned.space/me) on statbus because I really wanted to find this dumb song that an admin played a while back, that I couldn't find anymore. I still haven't found it, but I did find this

[![screenshot_20190305_131352](https://user-images.githubusercontent.com/5211576/53827193-87cc0200-3f48-11e9-87ce-bfdb58d73745.png)](https://sb.atlantaned.space/rounds/82772/played_url)

So I thought "hm, I know that the admin midi definitely does not work like that", so I checked the logs of that round

[![cobby_struggling_with_memes](https://user-images.githubusercontent.com/5211576/53827326-c6fa5300-3f48-11e9-8aa4-d6171c5fb454.png)](https://tgstation13.org/parsed-logs/basil/data/logs/2018/01/30/round-82772/game.html)

Well @ExcessiveUseOfCobblestone, now it does!

Also, later on in the PR's life, I found that youtube-dl parses end_time too, so I supported that cause it's cool.

:cl: JJRcop
admin: Play Internet Sound now respects start (like &t=xxx or &start=xxx) and end times (like &end=xxx). This allows admins to choose a snippet of media to play, like for example pick a song out of an hour long album compilation.
/:cl:

JJRcop committed:
Play Internet Sound respects start time
For example, adding &t=400 to youtube links.
Doesn't try to parse the start time, youtube-dl does that for us already. Will probably work on most sites not just youtube.

ExcessiveUseOfCobblestone commented:
lol

ExcessiveUseOfCobblestone commented:
wonder what song it was

subject217 commented:
you can click on the "screencap" cobby it's actually a link

ExcessiveUseOfCobblestone commented:
LOL fighting the fire boss with fire

ninjanomnom commented on code/modules/admin/verbs/playsound.dm:
```diff
@@ -67,7 +67,7 @@
 		var/web_sound_url = ""
 		var/stop_web_sounds = FALSE
-		var/pitch
+		var/list/music_extra_data = list("pitch", "seek")
```
Just start it off as an empty list, you can assign to nonexistent keys without issue and trying to access those same keys will return null just fine

JJRcop replied:
Thanks, I am already aware of this. The intent is to let people know what keys are there to be used easily. With that in mind, do you still want it to be an empty list?

JJRcop replied:
This issue is still valid in my most recent commit, I simply added "end" to the list.

ninjanomnom commented on code/modules/goonchat/browserOutput.dm:
```diff
@@ -115,12 +115,15 @@ GLOBAL_DATUM_INIT(iconCache, /savefile, new("tmp/iconCache.sav")) //Cache of ico
-	if(pitch)
-		music_data["musicRate"] = pitch
+
+	if(extra_data && extra_data.len)
```
```suggestion
	if(extra_data?.len)
```

SouDescolado replied:
@ninjanomnom This always works? Like it checks if it is not null and then checks the "subvalue"?

ExcessiveUseOfCobblestone replied:
it's equivalent

JJRcop replied:
That's cool! I'll implement that right away

JJRcop committed:
[Play Internet Sound] now supports end time too

Switched event handler for start to loadeddata from canplay

loadeddata fires when the first frame is available
this is quicker and more consistent than canplay, which tries
to predict at least a few moments of playability being
available before it fires

JJRcop committed:
[Internet Sound] music_extra_data now starts null

JJRcop commented:
Now it supports end time too! Normally youtube doesn't do end time on the website, it just ignores the tag, which is &end. However, youtube-dl parses it anyway and we can use it!
